backend/gsr_booking/views.py

```diff
@@ -275,3 +283,94 @@ def get(self, request):
                 request.user, request.user.booking_groups.filter(name="Penn Labs").first()
             )
         )
+
+
+class CreateShareCode(APIView):
+    """
+    Creates or retrieves a share code for a GSR booking
+    """
+
+    permission_classes = [IsAuthenticated]
+
+    def post(self, request):
+        booking_id = request.data.get("booking_id")
+
+        if not booking_id:
+            return Response({"error": "booking_id is required"}, status=400)
+
+        booking = GSRBooking.objects.filter(id=booking_id).first()
+        if booking is None:
+            return Response({"error": "Booking not found"}, status=404)
+
+        owner = booking.reservation.creator if booking.reservation else booking.user
+        if owner != request.user and not request.user.is_staff:
+            return Response(
+                {"error": "You do not have permission to share this booking"},
+                status=403,
+            )
+
+        try:
+            existing_code = GSRShareCode.objects.filter(booking=booking).first()
+
+            if existing_code and existing_code.is_valid():
+                serializer = GSRShareCodeSerializer(existing_code)
+                return Response(serializer.data)
+
+            share_code = GSRShareCode.objects.create(
+                code=GSRShareCode.generate_code(),
+                booking=booking,
+                owner=request.user,
+            )
+
+            serializer = GSRShareCodeSerializer(share_code)
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+        except APIError as e:
+            return Response({"error": str(e)}, status=400)
+
+
+class ViewSharedBooking(APIView):
+    """
+    View a booking using a share code, no authentication is needed
+    """
+
+    permission_classes = [AllowAny]
+
+    def get(self, request, code):
+        try:
+            share_code = GSRShareCode.objects.filter(code=code).first()
+
+            if share_code is None:
+                return Response({"error": "Invalid share code"}, status=404)
+
+            if not share_code.is_valid():
```

Comment:
Not 100% sure but if it's possible to abstract the validity stuff to the serializer, that could be better.